src/idp/sp_extractor.rs

```diff
@@ -0,0 +1,90 @@
+use crate::metadata::{EntityDescriptor};
+use super::error::Error;
+use crate::crypto;
+use std::str::FromStr;
+
+pub struct SPMetadataExtractor(EntityDescriptor);
+
+pub struct RequiredAttribute {
+    pub name: String,
+    pub format: Option<String>,
+}
+
+pub struct Acs {
+    pub bind_type: BindType,
+    pub url: String,
+}
+
+pub enum BindType {
+    Post,
+}
+
+impl SPMetadataExtractor {
+    pub fn try_from_xml(xml: &str) -> Result<Self, Box<dyn std::error::Error>> {
+        Ok(Self(EntityDescriptor::from_str(xml)?))
```

Comment:
```suggestion
        Ok(Self(xml.parse()?))
```
`FromStr` is implemented for `EntityDescriptor`, so you can just use parse.